typing/typeopt.ml

```diff
@@ -179,194 +189,330 @@ let bigarray_type_kind_and_layout env typ =
+      | [{ld_type}] ->
+        value_kind env ~loc ~visited ~depth ~num_nodes_visited ld_type
+      | [] | _ :: _ :: _ -> assert false
+    end
+  | _ -> begin
+      let is_mutable, num_nodes_visited, _, kinds =
+        List.fold_left
+          (fun (is_mutable, num_nodes_visited, idx, kinds)
+            (label:Types.label_declaration) ->
+            let num_nodes_visited = num_nodes_visited + 1 in
+            let num_nodes_visited, kinds, field_mutable =
+              (* CR layouts v2: when we add other layouts, we'll need to check
+                 here that we aren't about to call value_kind on a different
+                 sort (we can get this info from the label.ld_layout).  For now
+                 we rely on the sanity check at the top of value_kind to rule
+                 out void. *)
+              let (num_nodes_visited, kind) =
+                value_kind env ~loc ~visited ~depth ~num_nodes_visited
+                  label.ld_type
+              in
+              (num_nodes_visited, kind :: kinds, label.ld_mutable)
             in
-            begin match result with
-            | None -> num_nodes_visited, Pgenval
-            | Some (num_nodes_visited, _, consts, _, non_consts) ->
-              match non_consts with
-              | [] -> assert false  (* See [List.for_all is_constant], above *)
-              | _::_ ->
-                num_nodes_visited, Pvariant { consts; non_consts }
-            end
-        | Type_record (labels, record_representation) ->
-          let depth = depth + 1 in
-          let (is_mutable, num_nodes_visited), fields =
-            List.fold_left_map
-              (fun (is_mutable, num_nodes_visited)
-                   (label:Types.label_declaration) ->
-                 let is_mutable =
-                   match label.ld_mutable with
-                   | Mutable -> true
-                   | Immutable -> is_mutable
-                 in
-                 let num_nodes_visited = num_nodes_visited + 1 in
-                 let num_nodes_visited, field =
-                   loop env ~visited ~depth ~num_nodes_visited label.ld_type
-                  in
-                 (is_mutable, num_nodes_visited), field)
-              (false, num_nodes_visited) labels
-          in
-          if is_mutable then
-            num_nodes_visited, Pgenval
-          else begin match record_representation with
-            | Record_regular ->
-              num_nodes_visited,
-                Pvariant { consts = []; non_consts = [0, fields] }
+            let is_mutable =
+              match field_mutable with
+              | Mutable -> true
+              | Immutable -> is_mutable
+            in
+            is_mutable, num_nodes_visited, idx+1, kinds)
+          (false, num_nodes_visited, 0, []) labels
+      in
+      let kinds = List.rev kinds in
+      match is_mutable, kinds with
+      | true, _ -> (num_nodes_visited, Pgenval)
+      | false, [] -> (num_nodes_visited, Pintval)
```

Comment:
I don't see this case in the old code... unless it's the `gc_always_ignorable` check (which it probably is).